src/data/rel.lean

```diff
@@ -0,0 +1,187 @@
+local infixr ` ∘ ` :=rel.comp
+
+lemma comp_assoc (r : rel α β) (s : rel β γ) (t : rel γ δ) :
+  (r ∘ s) ∘ t = r ∘ s ∘ t :=
+begin
+  unfold comp, ext x w, split,
+  { rintros ⟨z, ⟨y, rxy, syz⟩, tzw⟩, exact ⟨y, rxy, z, syz, tzw⟩ },
+  rintros ⟨y, rxy, z, syz, tzw⟩, exact ⟨z, ⟨y, rxy, syz⟩, tzw⟩
+end
+
+@[simp]
+lemma comp_right_id (r : rel α β) : r ∘ @eq β = r :=
+by { unfold comp, ext y, simp }
+
+@[simp]
+lemma comp_left_id (r : rel α β) : @eq α ∘ r = r :=
+by { unfold comp, ext x, simp }
+
+lemma inv_id : inv (@eq α) = @eq α :=
+by { ext x y, split; apply eq.symm }
+
+lemma inv_comp (r : rel α β) (s : rel β γ) : inv (r ∘ s) = inv s ∘ inv r :=
+by { ext x z, simp [comp, inv, flip, and.comm] }
+
+def image (s : set α) : set β := {y | ∃ x ∈ s, r x y}
+
+lemma mem_image (y : β) (s : set α) : y ∈ image r s ↔ ∃ x ∈ s, r x y :=
+iff.refl _
+
+lemma image_mono {s t : set α} (h : s ⊆ t) : r.image s ⊆ r.image t :=
+assume y ⟨x, xs, rxy⟩, ⟨x, h xs, rxy⟩
+
+lemma image_inter (s t : set α) : r.image (s ∩ t) ⊆ r.image s ∩ r.image t :=
+assume y ⟨x, ⟨xs, xt⟩, rxy⟩, ⟨⟨x, xs, rxy⟩, ⟨x, xt, rxy⟩⟩
+
+lemma image_union (s t : set α) : r.image (s ∪ t) = r.image s ∪ r.image t :=
+set.eq_of_subset_of_subset
+  (λ y ⟨x, xst, rxy⟩, 
+    begin 
+      cases xst with xs xt, 
+      { left, exact ⟨x, xs, rxy⟩ },
+      right, exact ⟨x, xt, rxy⟩ 
+    end)
+  (λ y ymem, 
+    begin 
+      rcases ymem with ⟨x, xs, rxy⟩ | ⟨x, xt, rxy⟩; existsi x,
+      { split, { left, exact xs }, exact rxy},
+      split, { right, exact xt }, exact rxy 
+    end)
+
+@[simp]
+lemma image_id (s : set α) : image (@eq α) s = s :=
+by { ext x, simp [mem_image] }
+
+lemma image_comp (s : rel β γ) (t : set α) : image (r ∘ s) t = image s (image r t) :=
+begin 
+  ext z, simp only [mem_image, comp], split,
+  { rintros ⟨x, xt, y, rxy, syz⟩, exact ⟨y, ⟨x, xt, rxy⟩, syz⟩ },
+  rintros ⟨y, ⟨x, xt, rxy⟩, syz⟩, exact ⟨x, xt, y, rxy, syz⟩
+end
+
+lemma image_univ : r.image set.univ = r.codom := by { ext y, simp [mem_image, codom] }
+
+def preimage (s : set β) : set α := image (inv r) s
+
+lemma mem_preimage (x : α) (s : set β) : x ∈ preimage r s ↔ ∃ y ∈ s, r x y :=
+iff.refl _
+
+lemma preimage_def (s : set β) : preimage r s = {x | ∃ y ∈ s, r x y} :=
+set.ext $ λ x, mem_preimage _ _ _
+
+lemma preimage_mono {s t : set β} (h : s ⊆ t) : r.preimage s ⊆ r.preimage t :=
+image_mono _ h
+
+lemma preimage_inter (s t : set β) : r.preimage (s ∩ t) ⊆ r.preimage s ∩ r.preimage t :=
+image_inter _ s t
+
+lemma preimage_union (s t : set β) : r.preimage (s ∪ t) = r.preimage s ∪ r.preimage t :=
+image_union _ s t
+
+lemma preimage_id (s : set α) : preimage (@eq α) s = s :=
+by simp only [preimage, inv_id, image_id]
+
+lemma preimage_comp (s : rel β γ) (t : set γ) : 
+  preimage (r ∘ s) t = preimage r (preimage s t) :=
+by simp only [preimage, inv_comp, image_comp]
+
+lemma preimage_univ : r.preimage set.univ = r.dom :=
+by { rw [preimage, image_univ, codom_inv] }
+
+def core (s : set β) := {x | ∀ y, r x y → y ∈ s}
```

Comment:
Is there a Galois connection between `core` and `image`?

Reply:
Good call:
``` lean
theorem image_subset_iff (s : set α) (t : set β) : image r s ⊆ t ↔ s ⊆ core r t :=
iff.intro
  (λ h x xs y rxy, h ⟨x, xs, rxy⟩)
  (λ h y ⟨x, xs, rxy⟩, h xs y rxy)

theorem core_preimage_gc : galois_connection (image r) (core r) :=
image_subset_iff _
```